src/Facets/GenericSwapFacetV3.sol

```diff
@@ -0,0 +1,515 @@
+        bytes32 _transactionId,
+        string calldata _integrator,
+        string calldata _referrer,
+        address payable _receiver,
+        uint256 _minAmountOut,
+        LibSwap.SwapData[] calldata _swapData
+    ) external payable {
+        _depositMultipleERC20Tokens(_swapData);
+        _executeSwaps(_swapData, _transactionId, _receiver);
+        _transferNativeTokensAndEmitEvent(
+            _transactionId,
+            _integrator,
+            _referrer,
+            _receiver,
+            _minAmountOut,
+            _swapData
+        );
+    }
+
+    /// @notice Performs multiple swaps in one transaction, starting with ERC20 and ending with ERC20
+    /// @param _transactionId the transaction id associated with the operation
+    /// @param _integrator the name of the integrator
+    /// @param _referrer the address of the referrer
+    /// @param _receiver the address to receive the swapped tokens into (also excess tokens)
+    /// @param _minAmountOut the minimum amount of the final asset to receive
+    /// @param _swapData an object containing swap related data to perform swaps before bridging
+    function swapTokensMultipleV3ERC20ToERC20(
+        bytes32 _transactionId,
+        string calldata _integrator,
+        string calldata _referrer,
+        address payable _receiver,
+        uint256 _minAmountOut,
+        LibSwap.SwapData[] calldata _swapData
+    ) external payable {
+        _depositMultipleERC20Tokens(_swapData);
+        _executeSwaps(_swapData, _transactionId, _receiver);
+        _transferERC20TokensAndEmitEvent(
+            _transactionId,
+            _integrator,
+            _referrer,
+            _receiver,
+            _minAmountOut,
+            _swapData
+        );
+    }
+
+    /// @notice Performs multiple swaps in one transaction, starting with native and ending with ERC20
+    /// @param _transactionId the transaction id associated with the operation
+    /// @param _integrator the name of the integrator
+    /// @param _referrer the address of the referrer
+    /// @param _receiver the address to receive the swapped tokens into (also excess tokens)
+    /// @param _minAmountOut the minimum amount of the final asset to receive
+    /// @param _swapData an object containing swap related data to perform swaps before bridging
+    function swapTokensMultipleV3NativeToERC20(
+        bytes32 _transactionId,
+        string calldata _integrator,
+        string calldata _referrer,
+        address payable _receiver,
+        uint256 _minAmountOut,
+        LibSwap.SwapData[] calldata _swapData
+    ) external payable {
+        _executeSwaps(_swapData, _transactionId, _receiver);
+        _transferERC20TokensAndEmitEvent(
+            _transactionId,
+            _integrator,
+            _referrer,
+            _receiver,
+            _minAmountOut,
+            _swapData
+        );
+    }
+
+    /// Private helper methods ///
+
+    function _depositMultipleERC20Tokens(
```

Comment:
Depositing multiple tokens is a very advanced case, I don't really think the optimized version does need to support it. Instead we could deposit only tokens of the first swap (if an optimization like this saves us more gas)

Reply:
@maxklenk 
Yes I thought about this, too and made several tests. 
Since we currently mostly have two entries in the LibSwap array, the gas savings would be tiny (~0,15%) but we are more future-proof with this version. Especially in the light of Permit2 coming up which will also enable gasless multiswaps. And Jumper team had interest in some sort of panic button functionality that swaps all tokens to stables, for which this would also be a good use case. 

I would leave it in there but am also happy to remove it if you have strong feelings about it.

Reply:
Thanks for the details, if the gas savings are that small we can keep it as it is.